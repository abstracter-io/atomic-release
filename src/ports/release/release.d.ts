interface Release {
  getChangelog(): Promise<string | null>;

  getVersions(): Promise<string[]>;

  getNextVersion(): Promise<string>;

  getPreviousVersion(): Promise<string>;

  getMentionedIssues(): Promise<Set<string>>;

  getChangelogByVersion(version: string): Promise<string | null>;
}

export { Release };
