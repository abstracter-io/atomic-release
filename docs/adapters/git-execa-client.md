# GitExecaClient

An implemention of the SDK [GitClient](../ports/git-client.md) interface using [execa](https://github.com/sindresorhus/execa).

### Options

###### Optional properties are denoted by *

Type: `object literal`

##### remote*

Type: `string`  
Default: `origin`

##### workingDirectory*

Type: `string`  
Default: `process.cwd()`

### Example

```js
const { GitExecaClient } = require("@abstracter/atomic-release/adapters/git-execa-client");

const gitClient = new GitExecaClient({ remote: "origin2", workingDirectory: "/some/absolute/path" });
```
