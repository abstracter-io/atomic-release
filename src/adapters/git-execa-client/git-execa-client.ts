import execa from "execa";
import semver from "semver";
import { GitClient, MergedTag, Commit } from "../../ports/git-client";

type GitClientOptions = {
  remote: string;
  workingDirectory: string;
};

class GitExecaClient implements GitClient {
  private readonly options: GitClientOptions;

  public constructor(options?: Partial<GitClientOptions>) {
    this.options = {
      remote: options?.remote ?? "origin",
      workingDirectory: options?.workingDirectory ?? process.cwd(),
    };
  }

  private cli(args: string[], options?: execa.Options) {
    return execa("git", args, {
      cwd: this.options.workingDirectory,
      ...options,
    });
  }

  async log(range: string, format: string): Promise<string[]> {
    const logs: string[] = [];
    const delimiter = ":++:";
    const subprocess = await this.cli(["log", range, `--pretty=format:${format}${delimiter}`]);

    for (const log of subprocess.stdout.split(delimiter)) {
      if (log.length) {
        logs.push(log.startsWith("\n") ? log.substr(1) : log);
      }
    }

    return logs;
  }

  async cliVersion(): Promise<string> {
    const { stdout } = await this.cli(["--version"]);

    return stdout.split(" ")[2];
  }

  async refHash(ref: string): Promise<string> {
    const { stdout } = await this.cli(["rev-parse", ref]);

    return stdout;
  }

  async refName(ref: string): Promise<string> {
    const { stdout } = await this.cli(["rev-parse", "--abbrev-ref", ref]);

    return stdout;
  }

  /**
   * @param range A git log range.
   *
   * @example
   *
   * const gitClient = new GitClient();
   *
   * // Find all commits since ref 1234 (non-inclusive)
   * const commitSince = await gitClient.commits("1234..");
   *
   * // Find all commits until 1234 (inclusive)
   * const commitsUntil = await gitClient.commits("1234");
   */
  async commits(range: string): Promise<Commit[]> {
    const commits: Commit[] = [];
    const delimiter = ":<>:";
    const tagPattern = /tag: (.*),?/gi;
    // @src https://git-scm.com/docs/pretty-formats
    const formats = ["%H", "%s", "%b", "%N", "%D", "%ct", "%an", "%ae", "%cn", "%ce"];
    const rawCommits = await this.log(range, formats.join(delimiter));

    for (const rawCommit of rawCommits) {
      const tags: string[] = [];
      const logEntry = rawCommit.split(delimiter);

      let match: RegExpExecArray | null = null;

      for (const ref of logEntry[4].split(",")) {
        while ((match = tagPattern.exec(ref))) {
          tags.push(match[1]);
        }
      }

      commits.push({
        hash: logEntry[0],

        subject: logEntry[1],

        body: logEntry[2],

        notes: logEntry[3] ?? null,

        committedTimestamp: parseInt(logEntry[5], 10) * 1000,

        author: {
          name: logEntry[6],
          email: logEntry[7],
        },

        committer: {
          name: logEntry[8],
          email: logEntry[9],
        },

        tags,
      });
    }

    return commits;
  }

  async mergedTags(ref: string): Promise<MergedTag[]> {
    const minCliVersion = "2.7.0";
    const version = await this.cliVersion();

    // The following git command requires git CLI version to be >= 2.7.0
    // https://stackoverflow.com/a/39084124/1614199
    if (semver.satisfies(version, `>= ${minCliVersion}`)) {
      const tags: MergedTag[] = [];
      const delimiter = ":++:";
      const { stdout } = await this.cli([
        "tag",
        `--merged=${ref}`,
        `--format=%(refname:strip=2)${delimiter}%(objectname)`,
      ]);

      for (const tag of stdout.split("\n")) {
        if (tag.length) {
          const [name, hash] = tag.split(delimiter);

          tags.push({
            hash,
            name,
          });
        }
      }

      return tags;
    }
    //

    throw new Error(`Git version >= ${minCliVersion} is required. Found ${version}.`);
  }

  async remoteTagHash(tagName: string): Promise<string | null> {
    const { stdout } = await this.cli(["ls-remote", this.options.remote, "-t", `refs/tags/${tagName}`]);

    if (stdout.length) {
      return stdout.split("\t")[0];
    }

    return null;
  }

  async remoteBranchHash(branchName?: string): Promise<string | null> {
    const branch = branchName ?? (await this.refName("HEAD"));
    const { stdout } = await this.cli(["ls-remote", this.options.remote, "-h", `refs/heads/${branch}`]);

    if (stdout.length) {
      return stdout.split("\t")[0];
    }

    return null;
  }
}

export { GitExecaClient, GitClientOptions, MergedTag };
