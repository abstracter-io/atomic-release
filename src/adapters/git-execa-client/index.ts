export * from "./git-execa-client";
