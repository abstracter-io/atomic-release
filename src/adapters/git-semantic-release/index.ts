export * from "./git-semantic-release";
