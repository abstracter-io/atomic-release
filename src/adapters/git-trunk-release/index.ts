export * from "./git-trunk-release";
