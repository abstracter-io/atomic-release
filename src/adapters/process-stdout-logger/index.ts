export * from "./process-stdout-logger";
