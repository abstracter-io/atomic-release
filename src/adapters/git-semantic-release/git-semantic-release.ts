import semver from "semver";
import getStream from "get-stream";
import intoStream from "into-stream";
import { ConventionalChangelogPreset } from "conventional-changelog-preset-loader";
import conventionalChangelogPreset from "conventional-changelog-conventionalcommits";
import conventionalChangelogWriter, { ChangelogWriterContext } from "conventional-changelog-writer";
import { ConventionalCommit, sync as syncConventionalCommitParser } from "conventional-commits-parser";

import { Logger } from "../../ports/logger";
import { memoize } from "../../utils/memoize";
import { Release } from "../../ports/release";
import { processStdoutLogger } from "../process-stdout-logger";
import { GitExecaClient, MergedTag } from "../git-execa-client";

type GitSemanticReleaseOptions = {
  logger?: Logger;

  gitClient?: GitExecaClient;

  remote?: string;

  initialVersion?: string;

  stableBranchName: string;

  workingDirectory?: string;

  preReleaseBranches?: {
    [branchName: string]: string;
  };

  rawConventionalCommits?: (range: string) => Promise<{ hash: string; raw: string }[]>;

  isReleaseCommit?: (commit: ConventionalCommit) => boolean;

  conventionalChangelogPreset?: ConventionalChangelogPreset;

  conventionalChangelogWriterContext: ChangelogWriterContext | null;
};

type Options = Required<GitSemanticReleaseOptions>;

const sortTags = (tags: MergedTag[]): MergedTag[] => {
  return tags.sort((a, b) => {
    return semver.rcompare(a.name, b.name);
  });
};

const inc = (version: string, type: string, preReleaseId?: string) => {
  let v: string | null;

  if (preReleaseId) {
    v = semver.inc(version, "prerelease", preReleaseId);
  }
  else {
    v = semver.inc(version, type as semver.ReleaseType);
  }

  if (v === null) {
    throw new Error(`semantic version is '${version}' is not valid`);
  }

  return v;
};

const clean = (version: string) => {
  const v = semver.clean(version);

  if (v === null) {
    throw new Error(`semantic version is '${version}' is not valid`);
  }

  return v;
}

const defaultOptions = async (options: GitSemanticReleaseOptions): Promise<Options> => {
  const stableBranchName = options.stableBranchName;
  const workingDirectory = options.workingDirectory ?? process.cwd();
  const remote = options.remote ?? "origin";
  const gitClient = options.gitClient ?? new GitExecaClient({
    remote,
    workingDirectory,
  });
  const isReleaseCommit = (commit: ConventionalCommit): boolean => {
    const type = commit.type;

    if (!Object.prototype.hasOwnProperty.call(commit, "type")) {
      throw new Error("Non supported conventional commit. Provide a custom filter.");
    }

    if (typeof type === "string") {
      return /feat|fix|perf/.test(type);
    }

    return false;
  };

  const rawConventionalCommits = async (range: string) => {
    const commits = await gitClient.commits(range);

    return commits.map((commit) => {
      const lines = [
        // subject
        `${commit.subject}`,

        // body
        `${commit.body}`,

        // extra fields
        "-hash-",
        `${commit.hash}`,

        "-gitTags-",
        `${commit.tags.join(",")}`,

        "-committerDate-",
        `${new Date(commit.committedTimestamp)}`,
      ];

      return {
        hash: commit.hash,
        raw: lines.join("\n"),
      };
    });
  };

  return {
    remote,
    gitClient,
    workingDirectory,
    stableBranchName,

    logger: options.logger ?? processStdoutLogger({ name: "GitSemanticRelease" }),
    initialVersion: options.initialVersion ?? "0.0.0",
    preReleaseBranches: options.preReleaseBranches ?? {},
    isReleaseCommit: options.isReleaseCommit ?? isReleaseCommit,
    rawConventionalCommits: options.rawConventionalCommits ?? rawConventionalCommits,
    conventionalChangelogPreset: options.conventionalChangelogPreset ?? (await conventionalChangelogPreset()),
    conventionalChangelogWriterContext: options.conventionalChangelogWriterContext ?? null,
  };
};

const gitSemanticRelease = async (options: GitSemanticReleaseOptions): Promise<Release> => {
  const memo = memoize();
  const { logger, ...opt } = await defaultOptions(options);
  const preset = opt.conventionalChangelogPreset;
  const gitClient = opt.gitClient;

  // Private methods
  // ================
  const getBranchName = () => {
    return memo("branch_name", async () => {
      return gitClient.refName("HEAD");
    });
  };

  const getMergedTags = async (): Promise<MergedTag[]> => {
    return memo("tags", async () => {
      const mergedHeadTags = await gitClient.mergedTags("HEAD");
      const preReleaseId = await getPreReleaseId();
      const stableTags: MergedTag[] = [];
      const branchTags: MergedTag[] = [];
      const filteredTags: MergedTag[] = [];

      for (const tag of mergedHeadTags) {
        const preReleaseComponents = semver.prerelease(tag.name) as string[] | null;
        const tagPreReleaseId = preReleaseComponents?.[0] ?? undefined;

        if (!semver.valid(tag.name)) {
          logger.debug(`Filtered tag '${tag.name}'. Tag name is not a valid semantic version`);

          filteredTags.push(tag);
        }
        //
        else if (tagPreReleaseId === undefined) {
          stableTags.push(tag);
        }
        //
        else if (preReleaseId && tagPreReleaseId === preReleaseId) {
          branchTags.push(tag);
        }
      }

      filteredTags.length && logger.info(`Filtered ${filteredTags.length} tags`);

      return [...sortTags(branchTags), ...sortTags(stableTags)];
    });
  };

  const getPreReleaseId = async (): Promise<string | undefined> => {
    return memo("pre_release_id", async () => {
      if (opt.stableBranchName) {
        const branchName = await getBranchName();
        const preReleaseId = opt.preReleaseBranches[branchName];

        if (branchName === opt.stableBranchName) {
          return undefined;
        }

        if (preReleaseId) {
          return preReleaseId;
        }
        //

        throw new Error(`Could not find pre release id for branch '${branchName}'`);
      }

      throw new Error("Stable branch name is missing");
    });
  };

  const parseCommit = async (rawConventionalCommit: string): Promise<ConventionalCommit> => {
    return syncConventionalCommitParser(rawConventionalCommit, preset.parserOpts);
  };

  const getConventionalCommits = async (): Promise<ConventionalCommit[]> => {
    return memo("conventional_commits", async () => {
      const tags = await getMergedTags();
      const until = await gitClient.refHash("HEAD");

      const since = tags[0]?.hash;
      const range = since ? `${since}..` : until;
      const commits = await opt.rawConventionalCommits(range);
      const parsedCommits = commits.map((commit) => {
        return parseCommit(commit.raw);
      });

      logger.info(`Retrieving commits ${since ? `since ${since}` : `until ${until}`}`);

      return Promise.all(parsedCommits);
    });
  };

  const getAllVersionsConventionalCommits = async (): Promise<{ [version: string]: ConventionalCommit[] }> => {
    return memo("versions_conventional_commits", async () => {
      const tags = await getMergedTags();
      const versionsConventionalCommits = {};

      if (tags.length) {
        const taken = new Set();
        const untilHash = tags[0].hash;
        const rawConventionalCommits = await opt.rawConventionalCommits(untilHash);
        const hashIndices = new Map<string, number>();

        for (let i = 0; i < rawConventionalCommits.length; i++) {
          hashIndices.set(rawConventionalCommits[i].hash, i);
        }

        for (let i = tags.length - 1; i > -1; i -= 1) {
          const tag = tags[i];
          const conventionalCommits: ConventionalCommit[] = [];

          let j = hashIndices.get(tag.hash);

          /* istanbul ignore if */
          if (j === undefined) {
            throw new Error("Can this happen??");
          }

          for (let l = rawConventionalCommits.length; j < l; j += 1) {
            const commit = rawConventionalCommits[j];

            if (!taken.has(commit.hash)) {
              // eslint-disable-next-line no-await-in-loop
              conventionalCommits.push(await parseCommit(commit.raw));

              taken.add(commit.hash);
            }
          }

          versionsConventionalCommits[clean(tag.name)] = conventionalCommits;
        }
      }

      return versionsConventionalCommits;
    });
  };

  const getChangelogWriterContext = (): ChangelogWriterContext => {
    if (opt.conventionalChangelogWriterContext) {
      return opt.conventionalChangelogWriterContext;
    }

    throw new Error("conventional changelog writer context is missing");
  };

  // Public methods
  // ==============
  const getVersions = async (): Promise<string[]> => {
    return memo("versions", async () => {
      const tags = await getMergedTags();

      return tags.map((tags) => {
        return clean(tags.name);
      });
    });
  };

  const getChangelog = async (): Promise<string> => {
    return memo("changelog", async () => {
      const nextVersion = await getNextVersion();
      const commits = await getConventionalCommits();
      const writerOpts = preset.writerOpts;
      const writerContext = {
        ...getChangelogWriterContext(),
        version: nextVersion,
      };
      const stream = intoStream.object(commits).pipe(conventionalChangelogWriter(writerContext, writerOpts));

      return getStream(stream);
    });
  };

  const getNextVersion = async (): Promise<string> => {
    return memo<Promise<string>>("next_version", async () => {
      const previousVersion = await getPreviousVersion();
      const conventionalCommits = await getConventionalCommits();
      const releaseCommits = conventionalCommits.filter(opt.isReleaseCommit);
      const totalFiltered = conventionalCommits.length - releaseCommits.length;

      logger.info(`Found ${conventionalCommits.length} new commits`);

      totalFiltered && logger.info(`Filtered ${totalFiltered} commit${totalFiltered > 1 ? "s" : ""}`);

      if (releaseCommits.length) {
        const preReleaseId = await getPreReleaseId();
        const bump = preset.recommendedBumpOpts.whatBump(releaseCommits);
        const type = { 0: "major", 1: "minor", 2: "patch" }[bump.level];
        const next = inc(previousVersion, type, preReleaseId);
        const name = `v${next}`;
        const hash = await gitClient.remoteTagHash(name);

        logger.info(bump.reason);

        if (hash) {
          logger.warn(`Version ${next} was already released. (tag: ${name})`);
          logger.warn(`You can fix this by branching from ${hash}`);

          throw new Error(`A tag for version '${next}' already exists (tag hash: ${hash})`);
        }

        return next;
      }

      return previousVersion;
    });
  };

  const getMentionedIssues = async (): Promise<Set<string>> => {
    return memo("mentioned_issues", async () => {
      const issues = new Set<string>();
      const commits = await getConventionalCommits();

      for (const commit of commits) {
        for (const reference of commit.references) {
          const issue = reference.issue;

          if (issue) {
            issues.add(issue);
          }
        }
      }

      return issues;
    });
  };

  const getPreviousVersion = async (): Promise<string> => {
    return memo("previous_version", async () => {
      const versions = await getVersions();

      if (versions.length) {
        return versions[0];
      }

      logger.info(`Could not find a previous version. Will use ${opt.initialVersion} as initial version`);

      if (!semver.valid(opt.initialVersion)) {
        throw new Error(`${opt.initialVersion} is not a semantic version`);
      }

      return opt.initialVersion;
    });
  };

  const getChangelogByVersion = async (version: string): Promise<string> => {
    return memo(`changelog_${version}`, async () => {
      const versionsConventionalCommits = await getAllVersionsConventionalCommits();
      const commits = versionsConventionalCommits[version];

      if (commits) {
        const writerOpts = preset.writerOpts;
        const writerContext = {
          ...getChangelogWriterContext(),
          version,
        };
        const stream = intoStream.object(commits).pipe(conventionalChangelogWriter(writerContext, writerOpts));

        return getStream(stream);
      }

      throw new Error(`Could not find version ${version} conventional commits`);
    });
  };

  return {
    getVersions,
    getChangelog,
    getNextVersion,
    getMentionedIssues,
    getPreviousVersion,
    getChangelogByVersion,
  };
};

export { gitSemanticRelease, GitSemanticReleaseOptions };
