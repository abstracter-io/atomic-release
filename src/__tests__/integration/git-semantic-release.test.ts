import conventionalCommitsParser from "conventional-commits-parser";
import conventionalChangelogPreset from "conventional-changelog-conventionalcommits";

import { Release } from "../../ports/release";
import { GitExecaClient } from "../../adapters/git-execa-client";
import { gitSemanticRelease } from "../../adapters/git-semantic-release";

import { spiedLogger } from "../stubs/spied-logger";

const LOGGER = spiedLogger();
const HASH = "c658ea3e060490dced90dfb34c018d88b8e797f9";
const STABLE_BRANCH_NAME = "main";
const PRE_RELEASE_BRANCH_NAME = "beta";

class GitClientStub extends GitExecaClient {
  constructor() {
    super({ workingDirectory: process.cwd() });
  }

  cliVersion = jest.fn();
  refHash = jest.fn();
  refName = jest.fn();
  commits = jest.fn();
  mergedTags = jest.fn();
  remoteTagHash = jest.fn();
  remoteBranchHash = jest.fn();
}

const releaseOptions = () => {
  return {
    logger: LOGGER,
    stableBranchName: STABLE_BRANCH_NAME,
    preReleaseBranches: {
      [PRE_RELEASE_BRANCH_NAME]: PRE_RELEASE_BRANCH_NAME,
    },
    conventionalChangelogWriterContext: {
      owner: "t",
      repository: "t",
      host: "https://github.com",
      repoUrl: "https://github.com/t/t",
      date: "2021-10-08",
    },
  };
};

const commit = () => {
  return {
    subject: "feat: ...",
    body: "",
    notes: "",
    author: {
      name: "",
      email: "",
    },
    committer: {
      name: "",
      email: "",
    },
    tags: [],
    hash: HASH,
    committedTimestamp: 1633686020134,
  };
};

describe("git semantic release", () => {
  let release: Release;
  let gitClient: jest.Mocked<GitExecaClient>;

  beforeEach(async () => {
    gitClient = new GitClientStub() as unknown as jest.Mocked<GitExecaClient>;

    release = await gitSemanticRelease({
      gitClient,
      ...releaseOptions(),
    });

    gitClient.cliVersion.mockImplementation(async () => "2.7.0");

    gitClient.refHash.mockImplementation(async () => HASH);

    gitClient.commits.mockImplementation(async () => [commit()]);

    gitClient.refName.mockImplementation(async () => STABLE_BRANCH_NAME);

    gitClient.mergedTags.mockImplementation(async () => {
      const name = "v0.1.0";
      const hash = HASH;

      return [{ name, hash }];
    });

    gitClient.remoteTagHash.mockImplementation(async () => null);

    gitClient.remoteBranchHash.mockImplementation(async () => HASH);
  });

  test("filter non release commits", async () => {
    const expectedCommit = commit();
    const isReleaseCommit = jest.fn();
    const release = await gitSemanticRelease({
      gitClient,
      isReleaseCommit,
      ...releaseOptions(),
    });

    isReleaseCommit.mockImplementation(() => {
      return false;
    });

    gitClient.commits.mockImplementation(async () => [expectedCommit]);

    await release.getNextVersion();

    expect(LOGGER.info).toBeCalledWith(`Filtered 1 commit`);
    expect(isReleaseCommit).toBeCalledTimes(1);
  });

  test("next version patch is bumped", async () => {
    const version = "0.1.0";
    const previousTag = { name: `v${version}`, hash: HASH };

    gitClient.refName.mockImplementation(async () => STABLE_BRANCH_NAME);

    gitClient.commits.mockImplementation(async () => {
      return [
        {
          ...commit(),
          subject: "fix: ...",
        },
      ];
    });

    gitClient.mergedTags.mockImplementation(async () => [previousTag]);

    expect(await release.getNextVersion()).toStrictEqual("0.1.1");

    expect(LOGGER.info).toBeCalledWith(`Found 1 new commits`);
  });

  test("next version minor is bumped", async () => {
    const version = "0.1.0";
    const previousTag = { name: `v${version}`, hash: HASH };

    gitClient.refName.mockImplementation(async () => STABLE_BRANCH_NAME);

    gitClient.commits.mockImplementation(async () => [commit()]);

    gitClient.mergedTags.mockImplementation(async () => [previousTag]);

    expect(await release.getNextVersion()).toStrictEqual("0.2.0");
  });

  test("next version major is bumped", async () => {
    const version = "0.1.0";
    const previousTag = { name: `v${version}`, hash: HASH };

    gitClient.refName.mockImplementation(async () => STABLE_BRANCH_NAME);

    gitClient.commits.mockImplementation(async () => {
      return [
        {
          ...commit(),
          subject: "feat!: ...",
        },
      ];
    });

    gitClient.mergedTags.mockImplementation(async () => [previousTag]);

    expect(await release.getNextVersion()).toStrictEqual("1.0.0");
  });

  test("lists tags as released versions", async () => {
    const version = "0.1.0";
    const tag = { name: `v${version}`, hash: HASH };

    gitClient.mergedTags.mockImplementation(async () => [tag]);

    expect(await release.getVersions()).toStrictEqual([version]);
    expect(gitClient.mergedTags).toBeCalledWith("HEAD");
  });

  test("release change log is generated", async () => {
    const release = await gitSemanticRelease({
      ...releaseOptions(),

      gitClient,

      rawConventionalCommits: async (range: string) => {
        const commits = await gitClient.commits(range);

        return commits.map((commit) => {
          const lines = [
            // subject
            `${commit.subject}`,

            // body
            `${commit.body}`,

            // extra fields
            "-hash-",
            `${commit.hash}`,

            "-gitTags-",
            `${commit.tags.join(",")}`,

            "-committerDate-",
            `${new Date(1633686020134)}`,
          ];

          return {
            hash: commit.hash,
            raw: lines.join("\n"),
          };
        });
      },
    });

    const changelog = await release.getChangelog();

    expect(changelog).toMatchSnapshot();
  });

  test("list issues mentioned in commits", async () => {
    const version = "0.1.0";
    const previousTag = { name: `v${version}`, hash: HASH };

    gitClient.refName.mockImplementation(async () => STABLE_BRANCH_NAME);

    gitClient.commits.mockImplementation(async () => {
      return [
        {
          ...commit(),
          subject: "feat!: ... closes #3, #46, #39",
        },
        { ...commit(), subject: "Merge pull request #999 from repo/branch" },
      ];
    });

    gitClient.mergedTags.mockImplementation(async () => [previousTag]);

    expect(await release.getMentionedIssues()).toStrictEqual(new Set(["3", "39", "46", "999"]));
  });

  test("previous version is latest stable", async () => {
    const expectedVersion = "2.0.0";

    gitClient.refName.mockImplementation(async () => PRE_RELEASE_BRANCH_NAME);

    gitClient.mergedTags.mockImplementation(async () => {
      return [{ name: `v${expectedVersion}`, hash: HASH }];
    });

    expect(await release.getPreviousVersion()).toStrictEqual(expectedVersion);
  });

  test("next version bump using pre release", async () => {
    const version = "0.1.0-beta.0";
    const previousTag = { name: `v${version}`, hash: HASH };

    gitClient.refName.mockImplementation(async () => PRE_RELEASE_BRANCH_NAME);

    gitClient.commits.mockImplementation(async () => {
      return [
        {
          ...commit(),
          subject: "feat!: ...",
        },
      ];
    });

    gitClient.mergedTags.mockImplementation(async () => [previousTag]);

    expect(await release.getNextVersion()).toStrictEqual("0.1.0-beta.1");
  });

  test("non semantic tag names are filtered", async () => {
    const tag = { name: `v2.0`, hash: HASH };

    gitClient.mergedTags.mockImplementation(async () => [tag]);

    expect(await release.getVersions()).toHaveLength(0);
    expect(LOGGER.debug).toBeCalledWith(`Filtered tag '${tag.name}'. Tag name is not a valid semantic version`);
  });

  test("previous release change log is generated", async () => {
    const versions = await release.getVersions();

    for (const version of versions) {
      const changelog = await release.getChangelogByVersion(version);

      expect(changelog).toMatchSnapshot();
    }
  });

  test("previous version fallback to initial version", async () => {
    const initialVersion = "1.1.1";

    release = await gitSemanticRelease({
      ...releaseOptions(),
      gitClient,
      initialVersion,
    });

    gitClient.refName.mockImplementation(async () => STABLE_BRANCH_NAME);

    gitClient.mergedTags.mockImplementation(async () => []);

    expect(await release.getPreviousVersion()).toStrictEqual(initialVersion);

    expect(LOGGER.info).toBeCalledWith(
      `Could not find a previous version. Will use ${initialVersion} as initial version`
    );
  });

  test("previous version is the highest released version", async () => {
    const tags = [
      { name: `v0.1.0`, hash: HASH },
      { name: `v0.2.0`, hash: HASH },
      { name: `v0.3.0`, hash: HASH },
    ];

    gitClient.refName.mockImplementation(async () => STABLE_BRANCH_NAME);

    gitClient.mergedTags.mockImplementation(async () => tags);

    expect(await release.getPreviousVersion()).toStrictEqual("0.3.0");
  });

  test("rawConventionalCommits is called with 'since' range", async () => {
    const rawConventionalCommits = jest.fn();
    const release = await gitSemanticRelease({
      ...releaseOptions(),

      gitClient,

      rawConventionalCommits,
    });

    rawConventionalCommits.mockImplementation(() => []);

    await release.getNextVersion();

    expect(rawConventionalCommits).toBeCalledWith(`${HASH}..`);
  });

  test("rawConventionalCommits is called with 'until' range", async () => {
    const rawConventionalCommits = jest.fn();
    const release = await gitSemanticRelease({
      ...releaseOptions(),

      gitClient,

      rawConventionalCommits,
    });

    rawConventionalCommits.mockImplementation(() => []);

    gitClient.mergedTags.mockImplementation(async () => []);

    await release.getNextVersion();

    expect(rawConventionalCommits).toBeCalledWith(HASH);
  });

  test("versions are derived from pre release and stable tags", async () => {
    const stableTag = { name: "v0.1.0", hash: HASH };
    const preReleaseTag = { name: `v0.1.1-${PRE_RELEASE_BRANCH_NAME}.0`, hash: HASH };

    gitClient.refName.mockImplementation(async () => PRE_RELEASE_BRANCH_NAME);

    gitClient.mergedTags.mockImplementation(async () => [
      preReleaseTag,
      stableTag,

      // This is a tag of another branch, it should not be included
      { name: `v0.1.1-next.0`, hash: HASH },
    ]);

    expect(await release.getVersions()).toStrictEqual([
      // fp
      preReleaseTag.name.slice(1),
      stableTag.name.slice(1),
    ]);
  });

  test("next version fails when stable branch name is missing", async () => {
    const release = await gitSemanticRelease({
      ...releaseOptions(),
      gitClient,
      stableBranchName: undefined as never,
    });
    const expectedError = new Error(`Stable branch name is missing`);

    expect(await release.getNextVersion().catch((e) => e)).toStrictEqual(expectedError);
  });

  test("previous version fails when initial version is invalid", async () => {
    const initialVersion = "v2.0";
    const release = await gitSemanticRelease({
      ...releaseOptions(),
      gitClient,
      initialVersion,
    });
    const expectedError = new Error(`${initialVersion} is not a semantic version`);
    gitClient.refName.mockImplementation(async () => STABLE_BRANCH_NAME);

    gitClient.mergedTags.mockImplementation(async () => []);

    expect(await release.getPreviousVersion().catch((e) => e)).toStrictEqual(expectedError);
  });

  test("generating changelog fails when writer context is missing", async () => {
    const release = await gitSemanticRelease({
      ...releaseOptions(),
      gitClient,
      conventionalChangelogWriterContext: null as never,
    });
    const expectedError = new Error(`conventional changelog writer context is missing`);

    expect(await release.getChangelog().catch((e) => e)).toStrictEqual(expectedError);
  });

  test("next version fails when a tag with the same version exists", async () => {
    const nextVersion = "1.1.0";
    const tag = { name: `v1.0.0`, hash: HASH };
    const release = await gitSemanticRelease({
      ...releaseOptions(),
      gitClient,
    });
    const expectedError = new Error(`A tag for version '${nextVersion}' already exists (tag hash: ${tag.hash})`);

    gitClient.mergedTags.mockImplementation(async () => [tag]);
    gitClient.remoteTagHash.mockImplementation(async () => tag.hash);

    expect(await release.getNextVersion().catch((e) => e)).toStrictEqual(expectedError);

    expect(LOGGER.warn).toBeCalledWith(`Version ${nextVersion} was already released. (tag: v${nextVersion})`);

    expect(LOGGER.warn).toBeCalledWith(`You can fix this by branching from ${tag.hash}`);
  });

  test("default release commits are 'perf', 'fix' or 'feat' / 'feature'", async () => {
    const version = "1.0.0";
    const commits = [
      { ...commit(), subject: "perf: ..." },
      { ...commit(), subject: "fix: ..." },
      { ...commit(), subject: "feat: ..." },
      { ...commit(), subject: "feature: ..." },
      { ...commit(), subject: "docs: ..." },
    ];
    const expectedVersions = ["1.0.1", "1.0.1", "1.1.0", "1.1.0", "1.0.0"];

    gitClient.mergedTags.mockImplementation(async () => [{ name: `v${version}`, hash: HASH }]);

    for (const commit of commits) {
      const release = await gitSemanticRelease({
        gitClient,
        ...releaseOptions(),
      });

      gitClient.commits.mockImplementation(async () => [commit]);

      expect(await release.getNextVersion()).toStrictEqual(expectedVersions.shift());
    }
  });

  test("generating version changelog fails when version does not exists", async () => {
    const version = "1.1.1";
    const expectedError = new Error(`Could not find version ${version} conventional commits`);

    expect(await release.getChangelogByVersion(version).catch((e) => e)).toStrictEqual(expectedError);
  });

  test("default release commits filter throws when commit is missing type", async () => {
    const parser = jest.spyOn(conventionalCommitsParser, "sync");
    const expectedError = new Error("Non supported conventional commit. Provide a custom filter.");

    parser.mockImplementationOnce((rawCommit) => {
      const conventionalCommit = conventionalCommitsParser.sync(rawCommit);

      delete conventionalCommit.type;

      return conventionalCommit;
    });

    expect(await release.getNextVersion().catch((e) => e)).toStrictEqual(expectedError);
  });

  test("next version fails when pre release branch is missing pre release id", async () => {
    const release = await gitSemanticRelease({
      ...releaseOptions(),
      gitClient,
      preReleaseBranches: {},
    });
    const branchName = PRE_RELEASE_BRANCH_NAME;
    const expectedError = new Error(`Could not find pre release id for branch '${branchName}'`);

    gitClient.refName.mockImplementation(async () => branchName);

    expect(await release.getNextVersion().catch((e) => e)).toStrictEqual(expectedError);
  });

  test("default 'isReleaseCommit' throws when commit does not have 'type' property", async () => {
    const preset = await conventionalChangelogPreset();
    const expectedError = new Error(`Non supported conventional commit. Provide a custom filter.`);

    preset.parserOpts.headerCorrespondence = ["b", "c", "d"];

    const release = await gitSemanticRelease({
      gitClient,
      conventionalChangelogPreset: preset,
      ...releaseOptions(),
    });

    expect(await release.getNextVersion().catch((e) => e)).toStrictEqual(expectedError);
  });
});
