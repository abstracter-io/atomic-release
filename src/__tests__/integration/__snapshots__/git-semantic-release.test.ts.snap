// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`git semantic release previous release change log is generated 1`] = `
"## 0.1.0 (2021-10-08)


### Features

* ... ([c658ea3](https://github.com/t/t/commit/c658ea3e060490dced90dfb34c018d88b8e797f9))

"
`;

exports[`git semantic release release change log is generated 1`] = `
"## 0.2.0 (2021-10-08)


### Features

* ... ([c658ea3](https://github.com/t/t/commit/c658ea3e060490dced90dfb34c018d88b8e797f9))

"
`;
